### DISCUSSION TITLE
Sending system messages to chat

### DISCUSSION THREAD

**Discussion by luv2code** [DeabLabs/cannoli#8] [Q&A] [answered]
I would like have a chat cannoli that passes a system message before it passes the note.

My cannoli looks like:
![image](https://github.com/DeabLabs/cannoli/assets/918935/08432f8b-5613-4914-a13e-233b6dcebc0a)

It uses the [Drunk Person](https://github.com/f/awesome-chatgpt-prompts#act-as-a-drunk-person) prompt in the awesome prompts project.

this doesn't really work:
![image](https://github.com/DeabLabs/cannoli/assets/918935/25d2fe63-9fdf-4718-9ecb-7bf6cbdfe3ae)

This is what it's supposed to be like:
![image](https://github.com/DeabLabs/cannoli/assets/918935/0de9563c-73ef-4285-be00-f9c6aa1a3240)
This is the same prompt fed into the same GPT-4 model as a system message.

I've tried a few different things; but this is as close to success as I've gotten. Any tips?

**Comment by luv2code** [comment]
I think I got it.
![image](https://github.com/DeabLabs/cannoli/assets/918935/8aac55bf-1af1-4377-964d-b80863927583)

I was thinking the combination of the purple node and the {{Note}} var was special; but really, they are two separate things. I just need to put the prompt in the same reference node with the {{Note}} var:
![image](https://github.com/DeabLabs/cannoli/assets/918935/ada4538e-a5eb-46e9-89ed-22525b44944a)

I don't think it's sending a system message but maybe that doesn't matter?

**Reply by blindmansion** [reply]
Hey! The way to do this is to have two nodes that send to the blank call node:
1. The NOTE reference node
2. The content node with the system message you'd like to send beforehand. The system message will always be added before the other messages.

Let me know if this doesn't work as expected.

**Reply by luv2code** [reply]
like so?
![image](https://github.com/DeabLabs/cannoli/assets/918935/059366de-a1a5-47f7-8df0-5a090202e43c)
It doesn't really work:
![image](https://github.com/DeabLabs/cannoli/assets/918935/b4cc496b-88e6-4c52-a0c0-4960589e8ffb)

**Reply by luv2code** [reply]
I tried with both a gray arrow and a green arrow from the system node to the call node. that doesn't seem to make a difference.

**Reply by luv2code** [reply]
Further testing reveals that it works if I delete the note and re-run the cannoli. It doesn't seem to pass the system message on the first run. But it does on subsequent runs. I feel like this might be a bug?

**Reply by luv2code** [reply]
I am confounded. this one worked correctly the first and every subsequent time.
![image](https://github.com/DeabLabs/cannoli/assets/918935/c0b7843e-0eab-4677-8526-90e429e0a7cc)

**Reply by luv2code** [reply]
I created this one the exact same way and it doesn't work.
![image](https://github.com/DeabLabs/cannoli/assets/918935/67a51e96-b960-408f-a2b3-b54afecea025)

**Reply by luv2code** [reply]
Here is what I would expect to see from gpt-4 with the same system msg and prompt:

> Oh man, I got you. How about Cannabis-infused guacamole? Not only does it taste amazing, but it has that unique twist to really set it apart. I mean, who wouldn't remember a party where they ate weed guacamole?
> 
> Here’s a simple recipe, man:
> 
> Ingredients:
> 
> 3 ripe avocados
> 1 lime juiced
> 1 teaspoon salt
> 1/2 cup diced onion
> 3 tablespoons chopped fresh cilantro
> 2 roma/plum tomatoes
> 1 teaspoon minced garlic
> 1 pinch ground cayenne pepper
> 1 tablespoon cannabis oil
> First step is to mash the avocados in a bowl. Afterwards, you add lime juice and salt. Mix in onion, cilantro, tomatoes, and garlic. Then blend in the cannabis oil and sprinkle with cayenne pepper.
> 
> The trick is in the cannabis oil - to make it, you first gotta decarboxylate your grown cannabis in oven for about 30-40 minutes at 240 degrees Fahrenheit. Once that's done, you combine it in a saucepan with coconut oil and heat it at low for 2-3 hours. But remember - low and slow is the key! This way it doesn't burn your cannabis, but it does extract all the important stuff out of it.
> 
> Just remember to advise your guests about the cannabis in it - you don't want anyone freaking out!

**Reply by blindmansion** [reply, marked as answer]
Looks like the node you're expecting to be the system message  is not purple, so it's being interpreted as a call node rather than a content node. If you make that node purple it should work as expected.

Also a better way to keep track of what the LLM is seeing is with logging arrows. If you have a logging arrow (orange) going from the call node to a content node (purple) then you can see all the messages being sent.

**Reply by luv2code** [reply]
I'm realizing that I need a cheatsheet or dictionary of some sort that labels all the different pieces. The college is very very helpful, but I am having a hard time using it as a reference. I feel like I have to parse too much text to figure out the names of things. Maybe we could create a cheatsheet canvas with all the entities and their names, and maybe links to the canvases in the college that provides further explanations/examples of use.

I have had a lot of trouble getting the logging arrows to work; but that is probably just because I don't fully understand the call node vs content node. The logging arrow always needs to go from a call node to a content node?

is the purple {{NOTE}} node a special type of content node called a reference node?

**Reply by blindmansion** [reply]
That's a great idea! I'll make one of those and let you know when it's updated.

That's correct, a logging arrow is an orange arrow going from a call node to a content node (purple). Also, I recently added a toggle to switch the colors that get interpreted as content and call nodes, I think that is a bit easier because content nodes make more sense as a default, feel free to give it a try (in settings under canvas preferences).

Yes, if a content node only has a note reference (NOTE, @variable, [[link]]), then it's a "reference node". Those nodes can be written to, and have some other special properties. There's a section about them in the college.
